include/range/v3/utility/common_iterator.hpp

```diff
@@ -27,6 +31,105 @@ namespace ranges
+            struct pair_construct_hack<std::pair<First, Second>, std::pair<Arg1, Arg2>>
+              : meta::and_<
+                    std::is_constructible<First, Arg1>,
+                    std::is_constructible<Second, Arg2>,
+                    meta::defer<Constructible, std::pair<First, Second>, std::pair<Arg1, Arg2>>>
+            {};
+
+            template<typename First, typename Second, typename Arg1, typename Arg2>
+            struct pair_construct_hack<std::pair<First, Second>, common_pair<Arg1, Arg2>>
+              : meta::and_<
+                    std::is_constructible<First, Arg1>,
+                    std::is_constructible<Second, Arg2>,
+                    meta::defer<Constructible, std::pair<First, Second>, std::pair<Arg1, Arg2>>>
+            {};
+
+            template<typename T, typename... Args>
+            using is_nothrow_constructible_hack = meta::and_<
+                pair_construct_hack<T, Args...>,
+                meta::defer<std::is_nothrow_constructible, T, Args...>>;
+#else
+            template<typename T, typename... Args>
+            using pair_construct_hack = Constructible<T, Args &&...>;
+
+            template<typename T, typename... Args>
+            using is_nothrow_constructible_hack = std::is_nothrow_constructible<T, Args...>;
+#endif
+
+            template<typename T>
+            class operator_arrow_proxy
+            {
+                T value_;
+            public:
+                template<typename U, CONCEPT_REQUIRES_(pair_construct_hack<T, U>())>
+                constexpr explicit operator_arrow_proxy(U&& u)
+                    noexcept(is_nothrow_constructible_hack<T, U>::value)
+                  : value_(std::forward<U>(u))
+                {}
+                T const *operator->() const noexcept
+                {
+                    return std::addressof(value_);
+                }
+            };
+
+            template <class, class = void>
+            struct HasMemberOperatorArrow : std::false_type {};
+            template <class T>
+            struct HasMemberOperatorArrow<T, meta::void_<
+                decltype(std::declval<T>().operator->())>>
+              : std::true_type
+            {};
+
+            template<typename I,
+                CONCEPT_REQUIRES_(std::is_pointer<I>::value ||
+                    HasMemberOperatorArrow<I>())>
+            constexpr I operator_arrow_(const I& i, int)
+            noexcept(std::is_nothrow_copy_constructible<I>::value)
+            {
+                return i;
+            }
+
+            template<typename I,
+                CONCEPT_REQUIRES_(std::is_reference<iterator_reference_t<I>>())>
+            auto operator_arrow_(const I& i, long)
+                noexcept(noexcept(*i)) ->
+                decltype(std::addressof(std::declval<iterator_reference_t<I>&>()))
+            {
+                auto && tmp = *i;
+                return std::addressof(tmp);
```

Comment:
Since it's almost entirely a backwards compat mechanism, `common_iterator` implements `operator->` when `*i` is a *glvalue*, unlike `move_iterator` which only implements `operator->` for *lvalue* `*i`.  `tmp` is an lvalue here whenever `*i` is a glvalue, as required by `std::addressof`.